### ISSUE TITLE
Window title report is not supported

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mintty/mintty#303]
```
What steps will reproduce the problem?
1. echo -e "\e[21t"

What is the expected output? What do you see instead?
Expected: current window title (embedded in control sequence)
Received: "l" (embedded in control sequence)
(As long as the proper title is not reported, it would be better to respond 
nothing.)

Please use labels and text to provide additional information.
termout.c:do_winop sends the dummy
winmain.c:win_copy_title seems to retrieve the desired information already, 
would just need to be converted from wchar
```

Original issue reported on code.google.com by `towom...@googlemail.com` on 17 Nov 2011 at 12:23

**Comment by reikred** [comment]
I'm also looking for a way to get the title of a mintty window (in cygwin) using a shell command. Right now the GUI toolbar.CopyTitle  operation followed by a paste does work, but requires mouse operations and cannot be scripted. But using echo and a control sequence still provides an empty output.

It seems like OP has looked at the code and has  an idea what mods are required to generate a string output...

(for xterms I use xtermcontrol --get-title to get the title, but looking for a solution that works for mintty, Besides, I cannot seem to find an xtermcontrol binary for cygwin).

**Comment by mintty** [comment]
What do you want to do with the title? Please check whether the title push/pop function is sufficient for you (see #304).
It is considered a security issue to provide the title; a tricky crafted text output to the terminal could actually use it to invoke an arbitrary command on behalf of the user. That's why xterm suppresses this function unless explicitly configured to provide it.

**Comment by reikred** [comment]
I use window titles to name history files and dirstack files that I save every several  minutes  (using tcsh 'periodic' functionality or similar mechanism). The idea is to be able to recover the state of my shell windows after a crash or a hang condition, or  even after a voluntary restart.

For xterms I use xtermcontrol --get-title. For mintty terminals I do not have such a program, so I tried to echo control sequences directly to get the title. If there was a program that got the title from the command line inside the mintty, that would work for me. It does not have to involve a control sequence and using echo, nor does it have to write to the tty. Would that be safer? Heck, maybe mintty could somehow export an environment variable that contains the title, and keep it up to date, and make it available to the child processes?

As I mentioned, the GUI Copy Title menu does what I want, but I cannot reach it from a shell...

I hope I am making sense.

**Comment by mintty** [comment]
It is not possible to change an environment variable of a child process. I see 3 options:
* Mintty could set an environment variable with the initial title before starting a shell. However, if the title gets changed, this cannot be detected. Would that be sufficient for your purpose?
* I could add a control sequence to invoke the menu function "Copy Title".
* I could add title reporting as in xterm, enabled by an option but disabled by default. Retrieving the title would then work the same way as in xterm, e.g. with that program xtermcontrol. I am still hesitating, though, because I don't like to get involved in security discussions.

**Comment by reikred** [comment]
I think option1 is somewhat limited, (no updates) but I would use it if no other option exists.

I like option2, but how would I get from Copy Title to Paste Title? I ask because I do not know how to execute a paste from within a shell script. Does mintty understand and react to echo "Control-V"? Or would paste be a security hole equivalent to the "cat somefile" hole that allowWindowOps creates? 

I also like option3, but I understand why you would rather avoid it. BTW, I cannot find an xtermcontrol binary for cygwin so I would probably have to use echo anyway.

In summary, I think a modification of option2 (control sequence that copies Title to variable) may be a solution. But the syntax of the variable assignment has to be programmable, too, since there may be different shells running. That could be a setting and/or command line option for mintty?

I have nagging feeling I'm not thinking about the overall problem the right way, that a better solution is right around the corner, somehow.

**Comment by reikred** [comment]
I also have a general question that seems related: When using control sequences, such as the one to get title, why would a terminal emulator (xterm , mintty, etc) place the response in the input buffer (stdin). Does it not make more sense to place the response in the output buffer (stdout). If written to stdout, there would be no security problem.  My ignorance is probably showing, maybe a tty does not distinguish between stdin and stdout, it is all the same.

**Comment by mintty** [comment]
Communication between terminal and application is bi-directional and the tty distinguishes the two directions well; the "output" of the terminal (i.e. what it generates) is the input of the application (what it receives). It is not possible and would not make much sense to fill the output buffer of an application from an external source.

**Comment by mintty** [comment]
I have implemented and uploaded a new escape sequence to invoke the "Copy Title" function.
You can try it by compiling the current archive yourself or wait for the next release.
Usage:
echo -en '\e7721;1\a'
cat /dev/clipboard
# or: cp /dev/clipboard my-title-file

**Comment by reikred** [comment]
I have compiled and tested, I found one problem in the above recipe. It seems to me that the control string above is missing the ] (right bracket) character right after the \e character. That is, the echo recipe should read  (I hope html does not eat the right bracket)

echo -en '\e]7721;1\a'''

Also, in tcsh, I had to leave out the -en option, **or**  alternatively specify to use /bin/echo rather than builtin echo. With those modifications, I am able to get a copy of the title to /dev/clipboard!

Thanks much. This is the first time I have had any success in compiling a cygwin binary, and I learned a few other things along the way, including the fact that I need a copy of cygwin1.dll in the same directory where the binary is, or else it will not run. Overall I would have to say I'm delighted by how this enhancement request was handled!

**Comment by mintty** [comment]
Copy Title function released in 2.2.1.

**Comment by Jim-Simpson** [comment]
I also need to retrieve the window title in mintty. How do I get version 2.2.1, and what is that the version number for?

**Comment by mintty** [comment]
Unclear question. 2.2.1 was released almost a year ago, current version is 2.6.2.
You can download for self-compilation from this repository, or use your package manager; what's your system (`uname -a`)? If you use old MSYS or a prepackaged distribution like Babun, you may be out of luck and have to compile yourself.

**Comment by Jim-Simpson** [comment]
I'm running cygwin,  and "uname -a" gives: CYGWIN_NT-6.2-WOW64 Lina 1.7.17(0.262/5/3) 2012-10-19 14:39 i686 Cygwin

I think I've downloaded and installed the newest version of cygwin, but I'm new to cygwin (but not to UNIX and LINUX), and I don't know how to look for the version of mintty itself. But running "mintty -V" in a mintty terminal window gives:
mintty 1.1.2
(C) 2012 Andy Koppe

It looks like I don't have the newest version of mintty. Can I get that separately from cygwin and build it in cygwin?

**Comment by mintty** [comment]
Your release of cygwin and mintty is very old.
Of course you can compile mintty yourself, provided you have the development environment installed (packages gcc-core and make). But you would normally update your installation with the cygwin `setup` program (https://cygwin.com/setup-x86.exe).

**Comment by Jim-Simpson** [comment]
That did it. I had installed a new 64-bit version of cygwin, but I now think it didn't really update anything. When I installed the 32-bit version, it took much longer, and now the Copy Title works! Thanks for the help.

**Comment by Jim-Simpson** [comment]
That did it! I had installed the 64-bit cygwin, but I don't think it installed properly. I installed the 323-bit version, which took much longer, and now I have the Copy Title function. Thanks for the help!

**Comment by Jim-Simpson** [comment]
That did it! I had installed the 64-bit cygwin, but I don't think it installed properly. I installed the 323-bit version, which took much longer, and now I have the Copy Title function. Thanks for the help!
